zebra-consensus/src/block/check.rs

```diff
@@ -21,14 +22,15 @@ pub fn is_coinbase_first(block: &Block) -> Result<(), BoxError> {
     let first = block
         .transactions
         .get(0)
-        .ok_or("block has no transactions")?;
+        .ok_or(TransactionError::NoTransactions)?;
```

Comment:
I think the reason that the `#[from] TransactionError` impl isn't being used is because the checks convert the errors into `BoxError` and erase the error type: they return a `Result<(), BoxError>`.

This means that a function like `is_coinbase_first(block: &Block) -> Result<(), BoxError>` declares that it can return any kind of error at all.  This defeats the goal of having a `BlockError` type that enumerates every possible failure.

To fix this, the functions in `block::check` should all be changed to return a `Result<_, BlockError>`.  Then they can only fail through one of the enumerated error variants.

Reply:
This is fixed at https://github.com/ZcashFoundation/zebra/pull/1070/commits/1bc2d06e1065afe2fca7dd358788a3cc8c0d81a2